bee-protocol/src/worker/transaction/hasher.rs

```diff
@@ -0,0 +1,188 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+// the License. You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+// an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and limitations under the License.
+
+use crate::{
+    message::{uncompress_transaction_bytes, Transaction as TransactionMessage},
+    protocol::Protocol,
+    worker::transaction::{HashCache, ProcessorWorkerEvent},
+};
+
+use bee_common::{shutdown_stream::ShutdownStream, worker::Error as WorkerError};
+use bee_crypto::ternary::{
+    sponge::{BatchHasher, CurlPRounds, BATCH_SIZE},
+    Hash,
+};
+use bee_network::EndpointId;
+use bee_ternary::{T1B1Buf, T5B1Buf, TritBuf, Trits, T5B1};
+use bee_transaction::bundled::BundledTransactionField;
+
+use bytemuck::cast_slice;
+use futures::{
+    channel::mpsc,
+    task::{Context, Poll},
+    Stream, StreamExt,
+};
+use log::{debug, info, warn};
+use pin_project::pin_project;
+
+use std::pin::Pin;
+
+// If a batch has less than this number of transactions, the regular CurlP hasher is used instead
+// of the batched one.
+const BATCH_SIZE_THRESHOLD: usize = 3;
+
+type Receiver = ShutdownStream<mpsc::UnboundedReceiver<HasherWorkerEvent>>;
+
+pub(crate) struct HasherWorkerEvent {
+    pub(crate) from: EndpointId,
+    pub(crate) transaction: TransactionMessage,
+}
+
+#[pin_project(project = HasherWorkerProj)]
+pub(crate) struct HasherWorker {
+    processor_worker: mpsc::UnboundedSender<ProcessorWorkerEvent>,
+    #[pin]
+    receiver: Receiver,
+    cache: HashCache,
+    hasher: BatchHasher,
+    events: Vec<HasherWorkerEvent>,
+}
+
+impl HasherWorker {
+    pub(crate) fn new(
+        processor_worker: mpsc::UnboundedSender<ProcessorWorkerEvent>,
+        cache_size: usize,
+        receiver: Receiver,
+    ) -> Self {
+        assert!(BATCH_SIZE_THRESHOLD <= BATCH_SIZE);
+        Self {
+            processor_worker,
+            receiver,
+            cache: HashCache::new(cache_size),
+            hasher: BatchHasher::new(8019, CurlPRounds::Rounds81),
+            events: Vec::with_capacity(BATCH_SIZE),
+        }
+    }
+
+    async fn trigger_hashing(&mut self, batch_size: usize) {
+        if batch_size < BATCH_SIZE_THRESHOLD {
+            let hashes = self.hasher.hash_unbatched();
+            Self::send_hashes(hashes, &mut self.events, &mut self.processor_worker).await
+        } else {
+            let hashes = self.hasher.hash_batched();
+            Self::send_hashes(hashes, &mut self.events, &mut self.processor_worker).await
+        }
+        // FIXME: we could store the fraction of times we use the batched hasher
+    }
+
+    async fn send_hashes(
```

Comment:
Doesn't need to be async ?

Reply:
yes, remnants of the bounded channels era